client/server/middleware/profiler.js

```diff
@@ -0,0 +1,72 @@
+const fs = require( 'fs' );
+const path = require( 'path' );
+const v8Profiler = require( 'v8-profiler-next' );
+
+/**
+ * This should be dynamically imported:
+ *
+ * if ( shouldProfile ) {
+ *   app.use ( require( 'calypso/server/middleware/profiler' )() )
+ * }
+ *
+ * to avoid importing v8-profiler-next in production environments.
+ */
+module.exports = () => {
+	let IS_PROFILING = false;
+
+	// Generates a CPU profile compatible with VS Code's viewer.
+	v8Profiler.setGenerateType( 1 );
+
+	const profilesRoot = path.resolve(
+		path.join( __dirname, '../../../profiles', new Date().toISOString() )
+	);
+	try {
+		fs.accessSync( profilesRoot );
+	} catch {
+		fs.mkdirSync( profilesRoot, { recursive: true } );
+	}
+
+	return ( req, res, next ) => {
+		// Avoid profiling certain requests (like for static files) and don't
+		// start profiling if we already are.
+		if (
+			IS_PROFILING ||
+			! req.originalUrl.startsWith( '/' ) ||
+			req.originalUrl.startsWith( '/calypso/' ) ||
+			req.originalUrl.startsWith( '/service-worker' ) ||
+			req.originalUrl.startsWith( '/nostats.js' ) ||
+			req.originalUrl.startsWith( '/version' )
+		) {
+			next();
+			return;
+		}
+		IS_PROFILING = true;
+
+		// Replace slash with underscore:
+		const profileName = req.originalUrl.replace( /\//g, '_' );
+
+		v8Profiler.startProfiling( profileName, true );
+
+		// Once the request finishes, save the profile data to a file.
+		res.on( 'close', () => {
+			IS_PROFILING = false;
+			const profile = v8Profiler.stopProfiling( profileName );
+
+			profile.export( ( error, result ) => {
+				if ( error ) {
+					console.error( error );
+					return;
+				}
+				fs.writeFile( `${ profilesRoot }/${ profileName }.cpuprofile`, result, ( err ) => {
```

Comment:
This actually makes sense -- I just chose `_` to replace `/` since `/` is also used by the filesystem. So visiting `/` outputs `_`, visiting `/themes/foo` outputs `_themes_foo`. So in your case, that's just the base URL of Calypso.

Let me know if you have any better ideas!

Reply:
Thanks for elaborating, I think it makes sense indeed 👍